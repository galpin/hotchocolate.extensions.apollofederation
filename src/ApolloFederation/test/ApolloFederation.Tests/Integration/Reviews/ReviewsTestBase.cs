using System.Threading.Tasks;
using Xunit;

namespace HotChocolate.Extensions.ApolloFederation.Integration.Reviews;

/// <summary>
/// https://github.com/apollographql/federation-demo/blob/master/services/reviews/index.js
/// </summary>
public abstract class ReviewsTestBase : IntegrationTestBase
{
    [Fact]
    public async Task Service_sdl()
    {
        await ExecuteAndMatchSnapshotAsync(@"
        {
            _service {
                sdl
            }
        }
        ");
    }

    [Fact]
    public async Task Entities_when_one()
    {
        await ExecuteAndMatchSnapshotAsync(@"
        {
            _entities(representations: [
                { __typename: ""User"", id: ""1"" }
            ]) {
                ...on User {
                    __typename
                    id
                    username
                }
            }
        }
        ");
    }

    [Fact]
    public async Task Entities_when_many()
    {
        await ExecuteAndMatchSnapshotAsync(@"
        {
            _entities(representations: [
                { __typename: ""User"", id: ""1"" },
                { __typename: ""User"", id: ""2"" }
            ]) {
                ...on User {
                    __typename
                    id
                    username
                }
            }
        }
        ");
    }

    [Fact]
    public async Task Entities_when_mixed()
    {
        await ExecuteAndMatchSnapshotAsync(@"
        {
            _entities(representations: [
                { __typename: ""User"", id: ""1"" },
                { __typename: ""Product"", upc: ""1"" },
                { __typename: ""Review"", id: ""3"" }
            ]) {
                ...on User {
                    __typename
                    id
                    username
                }
                ...on Product {
                    __typename
                    upc
                }
                ...on Review {
                    __typename
                    id
                }
            }
        }
        ");
    }

    [Fact]
    public async Task Entities_when_not_found()
    {
        await ExecuteAndMatchSnapshotAsync(@"
        {
            _entities(representations: [
                { __typename: ""NotUser"" }
             ]) {
                ...on User {
                    __typename
                }
            }
        }
        ");
    }

    [Fact]
    public async Task Entities_when_typename_missing()
    {
        await ExecuteAndMatchSnapshotAsync(@"
        {
            _entities(representations: [
                { id: ""1"" }
             ]) {
                ...on User {
                    __typename
                }
            }
        }
        ");
    }

    [Fact]
    public async Task Entities_when_typename_not_string()
    {
        await ExecuteAndMatchSnapshotAsync(@"
        {
            _entities(representations: [
                { __typename: 42 }
             ]) {
                ...on User {
                    __typename
                }
            }
        }
        ");
    }
}