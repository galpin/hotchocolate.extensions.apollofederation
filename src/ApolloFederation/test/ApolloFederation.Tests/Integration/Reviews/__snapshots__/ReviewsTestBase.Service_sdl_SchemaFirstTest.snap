{
  "data": {
    "_service": {
      "sdl": "type Product @extends @key(fields: \u0022upc\u0022) {\n  upc: String! @external\n  reviews: [Review]\n}\n\ntype Review @key(fields: \u0022id\u0022) {\n  id: ID!\n  body: String\n  author: User @provides(fields: \u0022username\u0022)\n  product: Product\n}\n\ntype User @extends @key(fields: \u0022id\u0022) {\n  id: ID! @external\n  username: String @external\n  reviews: [Review]\n}"
    }
  }
}
